lib/fizzy/stack.hpp

```diff
@@ -58,32 +58,54 @@ class Stack : public std::vector<T>
 
 class OperandStack
 {
+    /// The size of the pre-allocated internal storage: 128 bytes.
+    static constexpr auto small_storage_size = 128 / sizeof(uint64_t);
+
     /// The pointer to the top item, or below the stack bottom if stack is empty.
     ///
     /// This pointer always alias m_storage, but it is kept as the first field
     /// because it is accessed the most. Therefore, it must be initialized
     /// in the constructor after the m_storage.
     uint64_t* m_top;
 
-    /// The storage for items.
-    std::unique_ptr<uint64_t[]> m_storage;
+    /// The bottom of the stack. Set in the constructor and never modified.
+    ///
+    /// TODO: This pointer is rarely used and may be removed.
+    ///       The value can be recomputed as (m_large_storage ? m_large_storage : m_small_storage).
+    ///       Moreover, methods using it may be replaced by onces not needing bottom of the stack.
+    uint64_t* m_bottom;
+
+    /// The pre-allocated internal storage.
+    uint64_t m_small_storage[small_storage_size];
+
+    /// The unbounded storage for items.
+    std::unique_ptr<uint64_t[]> m_large_storage;
```

Comment:
could be `std::array`

Reply:
Oh, I put comment on the wrong line, sorry, I meant it for the stack storage above

Reply:
I guessed that later.

We only take the pointer of `m_small_storage` once in constructor. So I don't think it is even worth to include `<array>` header as we are not using any of `std::array` features.